### PR TITLE
[master] Zil 4707 implement eth api calls in acceptance test framework

### PR DESCRIPTION
**Pull request by moboware**
## Description
To keep the PR small, I would like the current set of implemented test to be reviewed.

This implements tests for a part of the eth rpc api calls and small changes in the zilliqa code to fix the outcome.

## Backward Compatibility

- [X] This is not a breaking change
- [ ] This is a breaking change

## Review Suggestion
Please look at the js test implementations if these are solid enough.

## Status
Not all test are implemented.

### Implementation

- [X] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: tests/EvmAcceptanceTests/hardhat.config.js [hunk 1]
```diff
@@ -5,26 +5,36 @@ require("@nomiclabs/hardhat-web3");
 module.exports = {
   solidity: "0.8.9",
   defaultNetwork: "ganache",
+  // defaultNetwork: "isolated_server",
   networks: {
     ganache: {
-      url: "http://127.0.0.1:7545",
+      url: "http://localhost:7545",
       chainId: 1337,
+      web3_clientVersion: "Ganache/v7.4.1/EthereumJS TestRPC/v7.4.1/ethereum-js",
```
Comment (its-saeed) [inline tests/EvmAcceptanceTests/hardhat.config.js:13]: can we change it camelCase? `web3ClientVersion`

File: src/libServer/LookupServer.cpp [hunk 2]
```diff
@@ -1182,16 +1182,23 @@ Json::Value LookupServer::GetEthBlockCommon(const TxBlock& txBlock,
                                                  includeFullTransactions);
 }
 
-Json::Value LookupServer::GetEthBalance(const std::string& address) {
-  const auto balanceStr = this->GetBalance(address, true)["balance"].asString();
+Json::Value LookupServer::GetEthBalance(const std::string& address,
+                                        const std::string& tag) {
+  if (tag == "latest" || tag == "earliest" || tag == "pending") {
```
Comment (valeryz) [inline src/libServer/LookupServer.cpp:1187]: Let's mark it with a TODO and file a ticket saying that this is a limited current implementation and we'll need to have a full implementation of this for pending and earliest and any other block.

File: src/libServer/LookupServer.cpp [hunk 3]
```diff
@@ -1723,17 +1730,21 @@ Json::Value LookupServer::GetBalance(const string& address, bool noThrow) {
 
 std::string LookupServer::GetWeb3ClientVersion() {
   LOG_MARKER();
-  return "to do implement web3 version string";
+
+  return "Zilliqa/v8.1.2";
```
Comment (valeryz) [inline src/libServer/LookupServer.cpp:1734]: Could well be v8.2 - the next candidate.

File: tests/EvmAcceptanceTests/test/rpc/web3_sha3.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+const helper = require('../../helper/GeneralHelper');
+assert = require('chai').assert;
+
+const METHOD = 'web3_sha3';
+
+describe("Calling " + METHOD, function () {
+
```
Comment (its-saeed) [inline tests/EvmAcceptanceTests/test/rpc/web3_sha3.js:7]: could you remove these unnecessary empty lines? in other tests as well

File: tests/EvmAcceptanceTests/test/rpc/eth_estimateGas.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+const helper = require('../../helper/GeneralHelper');
+assert = require('chai').assert;
+
+const METHOD = 'eth_estimateGas';
+
+describe("Calling " + METHOD, function () {
+
+  it("should return the estimated gas as calculated over the transaction provided", async function () {
+    const estimatedGas = 2000000000
```
Comment (valeryz) [inline tests/EvmAcceptanceTests/test/rpc/eth_estimateGas.js:9]: Estimated gas of 2000000000 seems weird,  looks more like gasPrice.   Typical values of estimated gas is tens of thousands for hundreds of thusand for complicated contracts.

Comment (valeryz) [reply tests/EvmAcceptanceTests/test/rpc/eth_estimateGas.js:9]: I mean, check if the API actually implements something sensible.

Also, this is not a full implementation of the API yet, it is more of a stub for now.

**Review by valeryz — APPROVED** [review]
LGTM, but please address the comments.

File: tests/EvmAcceptanceTests/start_ganache.sh [hunk 6]
```diff
@@ -0,0 +1,2 @@
+clear
```
Comment (its-saeed) [inline tests/EvmAcceptanceTests/start_ganache.sh:1]: can you create an issue for porting this script to a [hardhat task](https://hardhat.org/hardhat-runner/docs/advanced/create-task) and assign it to me?

Comment (its-saeed) [reply tests/EvmAcceptanceTests/start_ganache.sh:1]: we can use hardhat-ganache plugin as well.

File: tests/EvmAcceptanceTests/test/rpc/eth_estimateGas.js (hunk 5 above)
Comment (moboware) [reply tests/EvmAcceptanceTests/test/rpc/eth_estimateGas.js:9]: Ganache come with defaults and that's what the test uses. The implementation is zilliqa is under construction, see ticket: https://zilliqa-jira.atlassian.net/browse/ZIL-4726

File: tests/EvmAcceptanceTests/start_ganache.sh (hunk 6 above)
Comment (moboware) [reply tests/EvmAcceptanceTests/start_ganache.sh:1]: https://zilliqa-jira.atlassian.net/browse/ZIL-4770

File: src/libServer/LookupServer.cpp (hunk 2 above)
Comment (moboware) [reply src/libServer/LookupServer.cpp:1187]: https://zilliqa-jira.atlassian.net/browse/ZIL-4771

File: tests/EvmAcceptanceTests/test/rpc/eth_gasPrice.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+const helper = require('../../helper/GeneralHelper');
+assert = require('chai').assert;
+
+const METHOD = 'eth_gasPrice';
+
+describe("Calling " + METHOD, function () {
+ 
+
+  it("should return the gasPrice as specified in the ethereum protocol", async function () {
+    const expectedGasPrice = 2000000000 // default ganache gas price in wei
```
[the remark this answers is not in the document]
Comment (moboware) [reply tests/EvmAcceptanceTests/test/rpc/eth_gasPrice.js:10]: This test will probably evolve

**Pull request merged**